### ISSUE TITLE
supervised learning using the pose info

### ISSUE THREAD

**Issue by YJonmo** [nianticlabs/monodepth2#237]
Hi there,

Thank you for this great work. I am more of a user of the deep learning approaches than a developer. Appreciate your help.

For my application, I have access to the groundtruth poses and I want to include it in the cost function with this hope to improve the pose and perhaps the depth estimation. Should I add my 'pose loss' to the current work (line 489) as follow?

abs_diff_pose = torch.abs(target_pose - pred_pose)
l1_loss_pose = abs_diff.mean(1, True)
loss += to_optimise.mean() + l1_loss_pose.mean()

I also notice that when using the 'pairs' then you are only using the first column of the pose network and the second column info are discarded:

outputs[("cam_T_cam", 0, f_i)] = transformation_from_parameters(
                        axisangle[:, 0], translation[:, 0], invert=(f_i < 0))

whereas when using the 'all' frames then all the pose outputs are used:
            for i, f_i in enumerate(self.opt.frame_ids[1:]):
                if f_i != "s":
                    outputs[("axisangle", 0, f_i)] = axisangle
                    outputs[("translation", 0, f_i)] = translation
                    outputs[("cam_T_cam", 0, f_i)] = transformation_from_parameters(
                        axisangle[:, i], translation[:, i])
Is that correct?

Regards,
Yaqub

**Comment by YJonmo** [comment]
I solved this thing by adding the groundtruth pose to the loss. However, my groundtruth pose is not aligned with the pose that the model['pose'] estimates. There should be some rotations applied to my groundtruth poses to make them aligned with the pose network. 

So now my question is, is there anyway to define the directions of the x, y, and z axis? if not then the only solution for me is to rotate my groundtruth coordinate to make them aligned (x: towards right, y: towards down, and z: towards the objects).

**Comment by mrharicot** [comment]
Hi,
You will need to rotate the ground truth poses such that it is using the same frame of reference as monodepth2 which is (as you said) right-handed with x-right, y-down, z-forward.

**Comment by YJonmo** [comment]
Hi Clement,

I just recorded a set of data with the coordinates being aligned with the computer vision standard (x rightward, y downward, z outward). 
Now my question is about the scaling of the translation. Since I want to use the groundtruth pose for estimation the loss, it is important to have the correct scale, otherwise the loss will be false. Is the self.K the only variable that I should change according to my camera to be able to get the correct scale out of the pose network (separate Resnet)?

https://github.com/nianticlabs/monodepth2/blob/101dd03f00e0d8e9bcd8a8d760f94bc2820f5549/datasets/kitti_dataset.py#L25

I can also see (as you said in the ReadMe) that the Base Line is set to 0.1:
https://github.com/nianticlabs/monodepth2/blob/101dd03f00e0d8e9bcd8a8d760f94bc2820f5549/datasets/mono_dataset.py#L196

In my case  as I am using an stereo endoscope the baseline is 1.5mm. So do I need to set the baseline to 0.0015?

**Comment by mrharicot** [comment]
Hi,

1. Yes you will need to edit the normalized intrinsics to match your camera parameters
2. I would recommend to keep the baseline to 0.1 and instead scale your ground truth translations by 0.0015 / 0.1. The reason being that it might be difficult to train if the baseline is different from 0.1 as it isn't very robust to the choice of this parameter.
3. Yes you would have to recompute the stereo scale factor in your case.

I hope this helps!

**Comment by YJonmo** [comment]
Thanks,

If I want to re-scale my groundtruth pose, then the only parameter that I should change in your code 'during the training' should be the K matrix. Is that right?

The other question is that my endoscope resolution is 400*400 but I cropped the images and then down sampled it to 256*256. Then what should be the K matrix?
        self.K = np.array([[330.0/400, 0, 200.0/400, 0],  
                           [0, 330.0/400, 200.0/400, 0],  
                           [0, 0, 1, 0],
                           [0, 0, 0, 1]], dtype=np.float32)

or 

        self.K = np.array([[330.0/256, 0, 200.0/256, 0],  
                           [0, 330.0/256, 200.0/256, 0], 
                           [0, 0, 1, 0],
                           [0, 0, 0, 1]], dtype=np.float32)


Cheers

**Comment by YJonmo** [comment]
By the way my pose GT are in mm, so I should divide them by 1000 first and then use the 0.0015 / 0.1 to rescale my GT pose? If I do so then my GT pose numbers become too small. Using only the 0.0015 /0.1 coefficient seems to be result in a closer .abs().mean() value to to the predicted poses.

**Comment by YJonmo** [comment]
I think I a figured it out, but I am still not sure about the K matrix and whether I should do the flipping or not:

    self.K = np.array([[330.0/400, 0, 200.0/400, 0],  
                       [0, 330.0/400, 200.0/400, 0],  
                       [0, 0, 1, 0],
                       [0, 0, 0, 1]], dtype=np.float32)

or

    self.K = np.array([[330.0/256, 0, 200.0/256, 0],  
                       [0, 330.0/256, 200.0/256, 0], 
                       [0, 0, 1, 0],
                       [0, 0, 0, 1]], dtype=np.float32)

**Comment by jilner** [comment]
How to get the ground truth pose?Thanks！

**Comment by YJonmo** [comment]
> How to get the ground truth pose?Thanks！

I already have my own groundtruth poses. I am not sure if Kitti data has pose info.
I might share code for supervised pose estimation if you wanted.

**Comment by YJonmo** [comment]
> How to get the ground truth pose?Thanks！

http://www.cvlibs.net/datasets/kitti/eval_odometry.php
